### ISSUE TITLE
Wrong decoding/encoding for 24bit unsigned integer

### ISSUE THREAD

**Issue by grammaprogramma** [eerimoq/asn1tools#103]
Hi, thank you for this great library. I think there is an issue when I use the C converter, in the following case. The ASN1 file declaration:
```
Counter24bit ::= INTEGER (0..16777215)	-- 24bit
```
is converted in the following C code 
```
static void packet_counter24bit_encode_inner(
    struct encoder_t *encoder_p,
    const struct lora_packet_lora_protocol_counter24bit_t *src_p)
{
    encoder_append_uint32(encoder_p, src_p->value);
}
```
instead of
```
static void packet_counter24bit_encode_inner(
    struct encoder_t *encoder_p,
    const struct lora_packet_lora_protocol_counter24bit_t *src_p)
{
    encoder_append_non_negative_binary_integer(
        encoder_p,
        (uint64_t)(src_p->value),
        24);
}
```
I think the issue is in `c/uper.py`, line 128, in the following test:
```
if (type_.number_of_bits % 8) ...
```
where we need to exclude number_of_bits == 24

Thank you!

**Comment by eerimoq** [comment]
Hi, glad you like it.

Ok, I've not worked on this project for a while but @Futsch1 has added some features to the C generator more recently. Maybe he can help out.

**Comment by Futsch1** [comment]
I will have a look!

**Comment by grammaprogramma** [comment]
Thank you very much!

 

Andrea Gramazio

founder & project manager

 




 <mailto:andrea.gramazio@lab3841.it> andrea.gramazio@lab3841.it |  <http://www.lab3841.it/> www.lab3841.it

via maestri del lavoro, 19 | 12100 cuneo | Italy

phone: +39 0171 290 720

mobile: +39 333 616 4170

 

Ai sensi e per gli effetti del D. Lgs. n. 196/2003 (Codice della Privacy) si precisa che questa e-mail è inviata unicamente ai destinatari sopra esposti con espressa diffida di leggerla, copiarla, usarla, senza autorizzazione. Se avete ricevuto questa copia per errore, vogliate distruggerla o contattarci immediatamente. Grazie per la Vostra collaborazione

This message may contain confidential and/or privileged information. If you are not the addressee or authorized to receive this for the addressee, you must not use, copy, disclose or take any action based on this message or any information herein. If you have received this message in error, please advise the sender immediately by reply e-mail and delete this message. Thank you for your cooperation

 

Da: Florian Fetz <notifications@github.com> 
Inviato: lunedì 11 gennaio 2021 21:04
A: eerimoq/asn1tools <asn1tools@noreply.github.com>
Cc: Andrea Gramazio <andrea.gramazio@lab3841.it>; Author <author@noreply.github.com>
Oggetto: Re: [eerimoq/asn1tools] Wrong decoding/encoding for 24bit unsigned integer (#103)

 

Closed #103 <https://github.com/eerimoq/asn1tools/issues/103>  via #104 <https://github.com/eerimoq/asn1tools/pull/104> .
